### DISCUSSION TITLE
Retrieving "RIGHT_ID" from tokens in `DependencyMatcher`

### DISCUSSION THREAD

**Discussion by felipemarineli** [explosion/spaCy#10823] [Help: Coding & Implementations] [answered]
Hi,

I am trying to retrieve the "RIGHT_ID" attribute for each token in a match using the `DependencyMatcher`. I could not figure out a way to do this. From the code snippet below, I would like to be able to output `anchor` and `author`. Is that possible?

```
    pattern = [
        # anchor token: founded
        {
            "RIGHT_ID": "anchor",
            "RIGHT_ATTRS": {"TEXT": {"REGEX": "[Bb]y"}}
        },
        {
            "LEFT_ID": "anchor",
            'REL_OP': ".*",
            "RIGHT_ID": "author",
            "RIGHT_ATTRS": {"ENT_TYPE": "PERSON"}
        }
    ]

    dependency_matcher.add("AuthorName", [pattern])
```

```
    for node in data:
        # apply NLP
        doc = nlp(input)  # Text by John Doe
        # apply RULER
        ruler_doc = ruler(doc)  # Text by John Doe
        # apply MATCHER on `doc` after RULER
        matches = matcher(ruler_doc)  # [(16787935407549884254, [1, 2])]
        ----- > print out RIGHT_ID of each match
```

Thank you in advance!

**Comment by adrianeboyd** [comment, marked as answer]
The list of matched tokens should correspond 1-to-1 to the pattern dicts, so `[1, 2]` means that the token with index 1 matched the first dict in the pattern and token with index 2 matched the second dict.

**Reply by felipemarineli** [reply]
Thank you for your reply!

I did manage to retrieve the `RIGHT_ID` in _most_ cases by retrieving the original pattern dictionaries and using the list of matched tokens. 

However, I do not seem to be able to retrieve the `RIGHT_ID` of the anchor token. Thus, in the example above, I can get `author`, but not `anchor`. This is fine for most cases, but in some cases, I am interested in the first token of the input `Doc`. For instance, if the input would be something like `John Doe, Correspondent`, I am interested in the correct matching of the whole thing (not any `PERSON` entity should be an author, but only the ones followed by "Correspondent"), but what I need to retrieve is the `author` bit (i.e. the `RIGHT_ID` of the anchor token).

Am I missing something?

**Reply by adrianeboyd** [reply]
Can you give a complete example that shows what you're trying to do? It's fine to use either manual annotation on the `doc` or a trained pipeline like `en_core_web_sm` (just specify which version so we can get the exact same annotation when we run the example on our end).

**Reply by felipemarineli** [reply]
Hi Ariane, thank you once again for your reply. I found the reason why I was not able to retrieve the `RIGHT_ID` of the anchor token, and it was not related to spaCy. I did manage now to do what I intended. Thank you :)
